Pyreverse autocolor

DudeNr33 opened the pull request:
<!--

Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to pylint in this document:
https://github.com/PyCQA/pylint/blob/master/doc/development_guide/contribute.rst#repository
-->

## Steps

- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description
Running ``pyreverse`` on large projects can lead to diagrams with low clarity. Classes or modules might be scattered across the diagram. 
By coloring the diagrams, modules or classes of the same package can be easily spotted visually. For an example, see issue #4488. 

This new feature is optional and can be activated by passing the ``--colorized`` option.
The granularity can be influenced by passing ``--max-color-depth``. 
Classes/modules from the standard library are always colored in grey. 
However, this seems to not work reliably, at least it did not work inside the tox test environments (I therefore had to delete the unittest for it). It worked great in my venv that I use for development.

This feature only works with dot output. I could not find a suitable program that correctly displays the vcg files.
As there were no unittests for the VCG output as well (are VCG graphs even still a thing? Google finds relatively few discussions for it...), I chose to not implement it for this output format.
I added a regression test which ensures that the VCG files produced with the current Pylint release and after this PR are the same.

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

## Related Issue

Closes #4488

Pierre-Sassoulas commented on pylint/pyreverse/writer.py:
```diff
@@ -123,7 +153,41 @@ def get_title(self, obj):
         return obj.title
 
-    def get_values(self, obj):
+    def get_style(self):
```
```suggestion
    def get_style(self) -> str:
```
Could you add typing on new functions please ? :)

DudeNr33 replied:
Will do!

Pierre-Sassoulas commented on pylint/pyreverse/writer.py:
```diff
@@ -184,7 +253,14 @@ def get_title(self, obj):
-    def get_values(self, obj):
+    def get_package_properties(self, obj):
+        """get label and shape for packages."""
+        return dict(
```
I think it would be better to use ``NamedTuple`` (or a child class) here, so we can have proper typing and not use dict everywhere.

DudeNr33 replied:
I agree that a NamedTuple would be the cleaner solution. However, the _current_ (as in: this PR) implementation of the ``DiagramWriter`` class and implementations of ``DotBackend`` and ``VCGPrinter`` do not allow for this. ``DotBackend`` and ``VCGPrinter`` rely on the data returned by the ``get_package_properties`` and ``get_class_properties`` methods, and the data structures they require are not the same.

That being said: I have already started working on the PlantUML backend, and while doing so I extracted a common interface for the ``Printer`` classes. Using a ``NamedTuple`` should be possible then.

Pierre-Sassoulas replied:
I agree with your other comment, it your implementation of plantuml make you upgrade the code with NamedTuple and proper typing it will be nice to rebase on it once it's merged.

Pierre-Sassoulas commented on tests/unittest_pyreverse_diadefs.py:
```diff
@@ -21,7 +21,7 @@
 import astroid
 import pytest
-from unittest_pyreverse_writer import Config, get_project
+from unittest_pyreverse_writer import _DEFAULTS, Config, get_project
```
Shouldn't this be defined in ``conftest.py`` and used as a pytest fixtures?

DudeNr33 replied:
Yes, that would be a good idea to keep the dependencies of the test modules as minimal as possible. I will do that.

Pierre-Sassoulas reviewed (changes requested):
Hey sorry for the delay in order to review. This looks nice and well tested, I think the pytest fixture were probably not used before (I see import from ``unittest_pyreverse_writer `` that would not happen if there were proper fixtures already). You can do a small refactor to move to pytest first if you have time for that or simply introduce a ``default_config`` fixture instead  of  the new ``_DEFAULT`` import.

DudeNr33 commented:
No problem, I fully understand that ``pyreverse`` is not the most important part of ``pylint``. ;) 

As I noted above, I already started working on the PlantUML implementation, and introducing ``NamedTuple`` in favor of dicts should be way easier there. My branch for PlantUML also already includes the changes of this PR.
My suggestion would be: 
We close this PR, and I will implement your review feedback in the PlantUML branch. 
I can then submit a new PR with both the coloring and the PlantUML support.

DudeNr33 commented:
@Pierre-Sassoulas quick question - I extracted the tests into a separate subdirectory and added a new ``conftest.py`` there.
Now ``mypy`` reports an error:
```
tests/conftest.py: error: Duplicate module named 'conftest' (also at 'tests/pyreverse/conftest.py')
```

As ``conftest.py`` modules may very well exist at different levels, I would go ahead and apply the suggested fix with the ``--exclude`` option.
But before I change anything of ``pylint``s static analysis rules, I'd like to double check if that's OK.

Pierre-Sassoulas commented:
It's a known problem with mypy, I don't know if you can exclude the file, if you can that's great, and I did not think of that because last time I put all my fixtures ``tests/conftests.py``, to make mypy happy 😄

DudeNr33 commented:
Yes, excluding works fine via the ``exclude`` option at the end of the ``mypy`` block in the ``pre-commit-config.yaml``:

```yaml
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.812
    hooks:
      - id: mypy
        name: mypy
        entry: mypy
        language: python
        types: [python]
        args: ["--ignore-missing-imports", "--scripts-are-modules"]
        require_serial: true
        additional_dependencies: []
        exclude: tests/functional/|tests/input|tests/extensions/data|tests/regrtest_data/|conftest.py|doc/|bin/
```

Pierre-Sassoulas commented:
Ha good to know, thanks ! Maybe we can ignore only some of them and keep analyzing ``test/conftests.py`` ? (Ie: Change ``|conftest.py|`` by ``|tests/pyreverse/conftest.py|`` ?)

DudeNr33 commented:
I see, being able to at least type check _one_ ``conftest.py`` would still be better than ignoring _all_ of them. 
I would not like to add every new ``conftest.py`` by hand, but I guess using ``tests(/.*)+/conftest.py`` should work.
This would still check the ``tests/conftest.py``, but ignore alle ``conftest.py`` which are in subdirectories of ``test``.

Pierre-Sassoulas commented:
Hey, just a heads up, we just merged #4551 so you'll probably want to rebase :)

DudeNr33 commented:
Thanks Pierre, I just submitted the new PR which also includes this feature as well as the requested changes in the review.
This PR can be closed.
